src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -759,3 +759,248 @@ - (void)invokeCommandWithEndpointID:(NSNumber *)endpointID
 }
 
 @end
+
+static BOOL MTRArrayIsStructurallyValid(NSArray * input, NSArray * structure)
```

Comment:
This could all easily live in a separate file dedicated to data validation rather than directly in MTRDevice.

Reply:
Done.